File: Python/ByTier/A/0923_A_Primal_Sport.py
# Problem: CF 923 A - Primal Sport
# https://codeforces.com/contest/923/problem/A

"""
Algorithm: Primal Sport
Approach:
- We are given X2, the state after two moves.
- We need to determine the minimum possible X0 that could lead to X2 after two moves.
- Each move involves selecting a prime p < current number and finding the smallest multiple of p >= current number.
- The key insight is to reverse-engineer the process:
  - For a given X2, find all possible X1 such that a valid prime operation from X1 to X2 is possible.
  - Then for each such X1, find possible X0 values by trying all primes < X1.

Time Complexity: O(N log log N + N) where N = 10^6
Space Complexity: O(N) for sieve and storing primes

"""

def sieve(MAX):
    sieve = [1] * MAX
    for i in range(2, MAX):
        if sieve[i] != 1:
            continue
        # Mark all multiples of i with i (smallest prime factor)
        for j in range(2 * i, MAX, i):
            sieve[j] = i
    return sieve


def main():
    MAX = int(1e6)
    N = MAX + 1
    # Generate smallest prime factor for all numbers up to 10^6
    primes = sieve(N)

    x2 = int(input())

    # Get the smallest prime factor of x2
    max_p = primes[x2]

    result = float("inf")
    # Iterate from x2 - max_p + 1 to x2 (possible values for x1)
    for x1 in range(x2 - max_p + 1, x2 + 1):
        # If x1 is prime, then x1 itself is the x0
        if primes[x1] == x1:
            result = min(result, x1)
            continue
        # Otherwise get the smallest prime factor of x1 (denoted as max_p)
        max_p = primes[x1]
        # x0 can be at most x1 - max_p + 1
        t = x1 - max_p + 1
        result = min(result, t)

    print(result)


main()


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/A/0923_A_Primal_Sport.cpp
// Problem: CF 923 A - Primal Sport
// https://codeforces.com/contest/923/problem/A

/*
 * Problem: Primal Sport
 * 
 * Algorithm/Technique: 
 *   - Sieve of Eratosthenes to precompute smallest prime factor (SPF) for each number up to n
 *   - Using SPF to compute a transformed value for each number
 *   - For each candidate starting number X0, we trace back the possible game states and find minimum
 * 
 * Time Complexity: O(n log log n) for sieve + O(n) for processing => O(n log log n)
 * Space Complexity: O(n) for storing smallest prime factors
 * 
 * The solution uses a reverse-engineering approach:
 *   - For each number x, compute the smallest prime factor (spf)
 *   - Transform each number as x -> x - spf(x) + 1
 *   - Then, for all numbers from the transformed value of n to n, find the minimum
 *   - The result corresponds to the minimal possible X0 that can lead to X2 = n after two moves
 */

#include <stdio.h>
#include <algorithm>

using namespace std;
const int maxn = 1e6 + 10;
int n, ans;
int f[maxn];

int main() {
  scanf("%d", &n);
  ans = n;
  // Sieve of Eratosthenes to compute smallest prime factor for each number
  for (int i = 2; i <= n; i++) {
    if (!f[i]) {  // i is prime
      for (int j = 2 * i; j <= n; j += i)
        f[j] = i;  // Mark smallest prime factor of j as i
    }
    f[i] = i - f[i] + 1;  // Transform the value: f[i] = i - spf(i) + 1
  }
  // Find minimum among all valid previous states
  for (int i = f[n]; i <= n; i++)
    ans = min(ans, f[i]);
  printf("%d\n", ans);
  return 0;
}


// https://github.com/VaHiX/CodeForces/